flow/record/adapter/splunk.py

```diff
@@ -23,21 +47,37 @@
         "source",
         "sourcetype",
         "tag",
+        "type",
     ]
 )
 
 
-def splunkify(record, tag=None):
+class Protocol(Enum):
+    HTTP = "http"
+    HTTPS = "https"
+    TCP = "tcp"
+
+
+class SourceType(Enum):
+    JSON = "json"
+    RECORDS = "records"
+
+
+def splunkify_key_value(record: Record, tag: Optional[str] = None) -> str:
     ret = []
 
-    ret.append(f'type="{record._desc.name}"')
+    ret.append(f'rdtype="{record._desc.name}"')
 
     if tag is None:
         ret.append("rdtag=None")
     else:
         ret.append(f'rdtag="{tag}"')
 
-    for field in record._desc.fields:
+    for field in record._desc.get_all_fields():
+        # Omit the _version field as the Splunk adapter has no reader support for deserialising records back.
+        if field == "_version":
```

Comment:
You don't do this in the JSON serialisation, is that intentional? And do the other `_` prefixed fields work fine?

Reply:
> You don't do this in the JSON serialisation, is that intentional? And do the other `_` prefixed fields not break anything? I vaguely remember being told not to do that, but I don't remember the reason.

Yes.

Reply:
Jokes aside, you were right that the JSON and records sourcetype were inconsistent in whether to include `_version`, it is now omitted for both. 

If I recall correctly, the `_` prefix should be fine as the reserved fields are guarded under the `SPLUNK_RESERVED_FIELDS` constant. I also vaguely recall warnings about breaking things, maybe you can ask Splunk people on your end? I haven't really touched anything Splunk related after the "Yes." incident.

<sup><sub>I miss you so much it hurts sometimes</sub></sup>

Reply:
I asked our Splunk experts internally and they recommended to add a custom prefix to `_` prefixed fields. This recommendation from Splunk can be found on this page: https://docs.splunk.com/Documentation/Splunk/7.3.1/Data/Configureindex-timefieldextraction (maybe nice to include this URL as a comment).

Reply:
I added a `rd_` prefix to the `_` prefixed fields (`_generated`, `_classification`), but while doing so ran into ambiguity for the `_source` field.  Sending data to the HEC supports overriding the `source` field while feeding data, and [Splunk's definition](https://docs.splunk.com/Splexicon:Source) of `source` seems to map perfectly on the usage of `_source` in flow.record. Therefore, the HEC writer already used `_source` for setting the `source` property.

However, the TCP forwarding method currently ignores `_source` in its entirety. No `source` value is set, and if a record has a field named `source` (e.g. `unix/user` from dissect.target), it is changed to `rd_source`. 

I think using `_source` field for Splunk's `source` would be nice, when available. Therefore I've added this to the TCP writer in the last commit and added a test to show how it works. One problem: I don't have access to the dissect technical addon for which the adapter was originally written, so I cannot test if this actually works.

These Splunk experts you speak of, do they have a Github account by any chance? It would be nice if we could wrap up this PR..

Reply:
I couldn't get them to comment on GitHub directly so you'll have to deal with me for a little longer. I hope you don't mind, otherwise we can talk it out over a nice dinner. 

I've relayed your proposal and come back with the conclusion that a) it's nice information to have available in Splunk, but b) that the `source` field is not the correct field for this. At least in the Dissect Splunk Technical Addon, the record descriptor type is used as `source` field, and that is a more appropriate use of that field.

We've come up with the following proposal (record field -> Splunk field):
- `source` -> `rd_source`
- `_source` -> `rd__source`

If you agree, it might be nice to also apply this to the other `_`-prefixed fields, where you currently remove the extra underscore. That way it provides a clear distinction between what fields were originally "private" flow record fields.

Reply:
Thanks for relaying, I've implemented the proposal. 

What's for dessert?

Reply:
Thanks, that looks good.

Dessert is all about the experience. It's not just what's on the plate but who's across from you. What about later, maybe we skip right to the sequel of what we finished last time? I promise it's a blockbuster.